[FIX] product: wrong serial number label format

ande-odoo opened the pull request:
Current behavior:
---
When printing a Lot/Serial Number, if the product name is too long, the content overlaps.

Cause of the issue:
---
Caused by: https://github.com/odoo/odoo/commit/95880f43c59c061f2a971ba8f41f9912139b798a
div was changed to span without changing the css

opw-3819349

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/159137).

nle-odoo reviewed (approved):
looks good to me

edited side note: since the element is a span, it seems that the only line that change something is the "font-size: 55%" (line-height and word-wrap don't work on "display: inline" (span default) elements, "white-space: normal; overflow: visible" are defaults)

ticodoo commented:
Hello, thanks for your fix but there are a few problems that aren't your fault:

1. the `span`s need to be changed back to `div` (can only be done in master due to potential existing xpaths, in stable you can wrap the existing spans with divs). I don't see a reason for why it was changed in the commit you linked, but my guess is they just changed it without understanding what they were doing. Because the 2 lines were changed to `span`, it's making the product name and the lot/sn name show on the same line which is WRONG
2. like @svs-odoo pointed out here https://github.com/odoo/odoo/pull/135739#issuecomment-1921212226, they also incorrectly changed the height for some reason here https://github.com/odoo/odoo/pull/135739/files#diff-fbc7f34c633189f45e17080184e52bb403bf631e36d8713cb0b2d27260629a3bR22 (i.e. they changed `height:19mm;` to `height:34mm;` in `report_lot_label`) so the label is the wrong size. Can you please change it back? There's already a fix PR in progress for making more than 1 of the lot labels print at a time so don't be surprised if your testing results in only 1 label per lot printed
3. this PR https://github.com/odoo/odoo/pull/122498 also incorrectly made it so the product name line wraps. I'm not sure if they original dev will fix it or not, but it's also incorrect, so the scss should be changed (see https://github.com/odoo/odoo/pull/122498#issuecomment-2033942467). You don't need to do anything for this comment, it's just an FYI because your commit message may not make sense anymore if it is changed

Also some commit message feedback:
- your commit header should say what your commit does, not what it is fixing, please see the [documentation](https://www.odoo.com/documentation/master/contributing/development/git_guidelines.html#commit-message-header): 
- for bugfixes, it's better to give the steps for how to reproduce the bug [there's an example here (except we don't include the affected versions)](https://github.com/odoo/odoo/wiki/Contributing#reporting-issues). I assume your bugfix is specifically for the datamatrix (gs1) use case because it doesn't make sense for the regular barcode use case.

ande-odoo commented:
Hi @ticodoo 

1: wrapped existing spans with divs
2: changed 34mm to 19mm
Commit feedback: fixed the title, forgot to include the steps, included them 

But now since the box is smaller (34 to 19mm) the issue appears again, if the name of the product is too long, the content gets out of the box, and overlaps on boxes below. 
What should we do to avoid that ? max-width or max-height ? with a max number of characters ?

ticodoo commented:
Hello @ande-odoo 

Indeed, it's the issue that was introduced by the other bugfix that I mentioned here

> 3. this PR [[FIX] product: product name overflow is hidden #122498](https://github.com/odoo/odoo/pull/122498) also incorrectly made it so the product name line wraps. I'm not sure if they original dev will fix it or not (see [[FIX] product: product name overflow is hidden #122498 (comment)](https://github.com/odoo/odoo/pull/122498#issuecomment-2033942467)). You don't need to do anything for this comment, it's just an FYI because your commit message may not make sense anymore if it is changed

So seems like the only solution is to revert that commit, but it needs to be done from v16 onwards :face_exhaling:

Also you may want to check that your changes doesn't mess up the formatting when the lot/sn has an expiration date due to the xpath's referring to the span in `product_expiry/report_lot_barcode.xml`

ande-odoo commented:
Hi @ticodoo 

I have reverted @momegahed 's commit (he doesn't work here anymore)

I have modified the view to make the label more readable, depending on the barcode type.
If it's a normal barcode, it displays the whole product name on one line, hiding what's beyond the box, to leave space for expiration dates.
And if it's GS1, I put the code on the first line, then the product name on the second, again leaving enough space for the expiration dates.

Concerning the version, I'll make a backport of this PR once it's merged to avoid a retarget

Please tell me what you think

ticodoo commented on addons/stock/report/report_lot_barcode.xml:
```diff
@@ -19,15 +19,29 @@
-                                    <span class="o_label_4x12" t-field="o.product_id.display_name" t-att-style="'width:22mm' if final_barcode else ''">Demo Product</span>
-                                    <span class="o_label_4x12" name="lot_name" t-field="o.name" t-att-style="'width:22mm' if final_barcode else ''">Demo Name</span>
+                                    <div class="o_label_4x12">
+                                        <t t-if="env.user.has_group('stock.group_stock_lot_print_gs1')">
```
Do you need this extra group check? The name splitting should apply to both gs1 and non-gs1 barcodes

ticodoo commented on addons/stock/report/report_lot_barcode.xml:
```diff
@@ -19,15 +19,29 @@
-                                    <span class="o_label_4x12" name="lot_name" t-field="o.name" t-att-style="'width:22mm' if final_barcode else ''">Demo Name</span>
+                                    <div class="o_label_4x12">
+                                        <t t-if="env.user.has_group('stock.group_stock_lot_print_gs1')">
+                                            <t t-set="code" t-value="o.product_id.default_code"/>
+                                            <t t-set="product_name" t-value="o.product_id.name"/>
```
I think this template is already complicated enough without extra variables to keep track of. Please change the logic of this back to the way it was before the change that removed it https://github.com/odoo/odoo/pull/135739/commits/77ed573bd26cf79d3bce919ab1ad45a9baa9a885#diff-fbc7f34c633189f45e17080184e52bb403bf631e36d8713cb0b2d27260629a3b

For reference, this is how the name was originally split: https://github.com/odoo/odoo/pull/122511/commits/9ffb7110a8b0f5f8aeb6418fd2c73338a69c7068

ticodoo commented on addons/product/static/src/scss/report_label_sheet.scss:
```diff
@@ -72,8 +72,8 @@
         padding:0;
         line-height:1;
         font-size:55%;
-        overflow:visible;
-        white-space:normal;
-        word-wrap: break-word;
```
This change is also in v16, so the same bug should occur there as well. It would be cleaner to do a PR that reverts the commit that added this in (i.e. with the [REV] tag + git revert [commit] command with an explanation as to why you're reverting it) targeting v16 and then either its fw-port adds in an extra commit with the `report_lot_barcode.xml` fix or open up a second PR that fixes `report_lot_barcode.xml`

ande-odoo commented:
As discussed, new PR in 16.0 reversing the faulty commit: https://github.com/odoo/odoo/pull/169226

ticodoo commented on addons/stock/report/report_lot_barcode.xml:
```diff
@@ -19,15 +19,25 @@
-                                    <span class="o_label_4x12" t-field="o.product_id.display_name" t-att-style="'width:22mm' if final_barcode else ''">Demo Product</span>
-                                    <span class="o_label_4x12" name="lot_name" t-field="o.name" t-att-style="'width:22mm' if final_barcode else ''">Demo Name</span>
+                                    <div class="o_label_4x12">
+                                        <t t-if="env.user.has_group('stock.group_stock_lot_print_gs1')">
```
Please delete this, it isn't necessary

ticodoo commented on addons/stock/report/report_lot_barcode.xml:
```diff
@@ -19,15 +19,25 @@
+                                            <span class="o_label_4x12" t-if="o.product_id.default_code" t-att-style="'width:22mm' if final_barcode else ''">[<t t-out="o.product_id.default_code"/>]</span><br />
+                                            <span class="o_label_4x12" t-field="o.product_id.name" t-att-style="'width:22mm' if final_barcode else ''">Demo Product</span>
+                                        </t>
+                                        <t t-else="">
+                                            <span class="o_label_4x12" t-field="o.product_id.display_name" t-att-style="'width:22mm' if final_barcode else ''">Demo Product</span>
+                                        </t>
```
drop this part

ticodoo reviewed (commented):
now that I look at the commit messages/PRs references again, it's supposed to be
```
[default code]
product name
```
for both the gs1 and regular barcode cases

ticodoo commented:
There were still a couple issues I noticed when testing, so I applied the final fixes so that we can get this done

before:
![image](https://github.com/user-attachments/assets/d50fc374-0f1e-462d-8f23-30a0f255e631)

after:
![image](https://github.com/user-attachments/assets/81f1b51a-4929-4048-a37a-5c9f98fdef27)

@robodoo r+
